Whole album matching and fingerprinting

ta264 opened the pull request:
#### Database Migration
YES (24, to remove old format MediaInfo data)

#### Description
As discussed on discord, adding fingerprinting and reworking the matching to be more like beets, taking the album as a whole.

#### Todos
- [x] Tests
- [x] UI to select release in manual import
- [x] Fix fingerprinting on windows and macOS

#### Issues Fixed
 - Fixes #512 
 - Fixes #268 
 - Fixes #296

Qstick commented:
Whatever file we are using and bundling for testing likely needs to be completely free, royalty-free, license-free, free to distribute, etc. etc. etc.. blah blah blah... :)

Not sure if `src/NzbDrone.Core.Test/Files/Media/nin.flac` qualifies?

ta264 commented:
`nin.flac` is a track from [here](https://en.m.wikipedia.org/wiki/Ghosts_I–IV) released under creative commons, so hopefully fine. There's a license in the tags but can put a license file in too if that helps.

Qstick commented:
Sounds good 👍

Qstick commented on src/NzbDrone.Core/MediaFiles/TrackImport/Aggregation/AggregationService.cs~:
```diff
@@ -0,0 +1,65 @@
+using System;
```
Stray file?

Qstick commented on src/NzbDrone.Core/Parser/Model/MediaInfoModel.cs:
```diff
@@ -0,0 +1,13 @@
+        public string AudioFormat { get; set; }
+        public int AudioBitrate { get; set; }
+        public int AudioChannels { get; set; }
+        public int SchemaRevision { get; set; }
```
Can we add Bits and SampleRate here. Taglib provides both I believe. Bits should be used to identify certain qualities (24bit Flac).

Qstick commented on src/NzbDrone.Core/Organizer/FileNameBuilder.cs:
```diff
@@ -315,74 +313,6 @@ private void AddQualityTokens(Dictionary<string, Func<TokenMatch, string>> token
-
-            tokenHandlers["{MediaInfo Simple}"] = m => $"{audioCodec}";
-
-            tokenHandlers["{MediaInfo Full}"] = m => $"{audioCodec}{mediaInfoAudioLanguages} {mediaInfoSubtitleLanguages}";
```
Should we keep the naming tokens in since we are still storing audio data, just from a different source?

ta264 replied:
Can do.  Will simplify to codec and channels.  The tricky bit is dealing with old MediaInfo entries in the database Vs new Taglib based ones.  There's no mechanism for updating them any more unless the track is actually reimported.  Or I could do a DB migration...

Qstick commented on src/NzbDrone.Core/Download/TrackedDownloads/TrackedDownloadService.cs:
```diff
@@ -186,7 +192,7 @@ private void LogItemChange(TrackedDownload trackedDownload, DownloadClientItem e
                 existingItem.CanMoveFiles != downloadItem.CanMoveFiles)
             {
-                _logger.Debug("Tracking '{0}:{1}': ClientState={2}{3} SonarrStage={4} Episode='{5}' OutputPath={6}.",
+                _logger.Debug("Tracking '{0}:{1}': ClientState={2}{3} SonarrStage={4} Album='{5}' OutputPath={6}.",
```
Can we go ahead and change this to `LidarrStage` as well?

Qstick commented on src/NzbDrone.Core/History/History.cs:
```diff
@@ -41,6 +41,8 @@ public enum HistoryEventType
-        TrackFileRenamed = 6
+        TrackFileRenamed = 6,
+        AlbumImportIncomplete = 7,
+        DownloadComplete = 8
```
Whats the difference in this and downloadFolderImported?

ta264 replied:
DownloadFolderImported should really be renamed to TrackFileImported i.e. it's logged for every track imported.  Previously one of these events was sufficient to mark a download as imported, but now there's a new event to mark that.

ta264 commented on frontend/src/Store/Actions/artistIndexActions.js:
```diff
@@ -76,8 +76,7 @@ export const defaultState = {
       label: 'Type',
       isSortable: true,
-      isVisible: true,
-      isModifiable: false
```
Oh that's a local UI change that has snuck in. I use the table view and it was annoying me that on my phone it insists on tellling me whether an artist is a group or an individual. I generally know or don't care and there isn't the real estate on my phone screen.  This change lets you disable that column.

ta264 commented on src/NzbDrone.Core/HealthCheck/Checks/MediaInfoDllCheck.cs:
```diff
@@ -1,24 +0,0 @@
-using System;
```
Good idea, will do

ta264 commented:
Will do some final testing and code review on this tomorrow but I think I'm pretty much happy with this now

Qstick commented:
@danielunderwood have you had a chance to test on macos yet?

Qstick commented on src/NzbDrone.Core/NzbDrone.Core.csproj:
```diff
@@ -110,8 +106,8 @@
-    <Reference Include="taglib-sharp, Version=2.1.0.0, Culture=neutral, PublicKeyToken=db62eba44689b5b0, processorArchitecture=MSIL">
-      <HintPath>..\packages\taglib.2.1.0.0\lib\taglib-sharp.dll</HintPath>
+    <Reference Include="taglib-sharp, Version=2.2.0.0, Culture=neutral, PublicKeyToken=db62eba44689b5b0, processorArchitecture=MSIL">
+      <HintPath>..\packages\TagLibSharp.2.2.0-beta/lib/netstandard2.0/taglib-sharp.dll</HintPath>
```
```suggestion
      <HintPath>..\packages\TagLibSharp.2.2.0-beta\lib\netstandard2.0\taglib-sharp.dll</HintPath>
```

Qstick commented on src/Lidarr.Api.V1/Artist/ArtistModule.cs:
```diff
@@ -245,17 +245,17 @@ public void Handle(TrackFileDeletedEvent message)
         public void Handle(ArtistUpdatedEvent message)
         {
-            BroadcastResourceChange(ModelAction.Updated, message.Artist.Id);
+            BroadcastResourceChange(ModelAction.Updated, message.Artist.ToResource());
```
This broadcast resource changes break updates to UI. Changing back to Id causes updates to come thru as expected. 

Ex. Import artist and sit on artist index page. At end of import final stats don't get updated correctly, However reverting to using function with `int Id` overlord causes resource to be retrieved fully and then sent to the UI.

This is because simply passing the `artist.toResource()` doesn't populate the entire resource (ex. statistics property  is left as `new ArtistStatisticsResource()`.) When using the `BroadcastResourceChange(ModelAction.Updated, message.Artist.Id)` overlord the function grabs the resource by using `GetResourceById` which in the artist case triggers the `GetArtist` function in ArtistModule.cs to return the resource. This function fully populates the resource (adds statistics, covers, etc.) and then broadcasts to the UI. Best to leave as is and use the Id.

ta264 replied:
Good catch

Qstick commented on src/Lidarr.Api.V1/Albums/AlbumModuleWithSignalR.cs:
```diff
@@ -138,12 +138,12 @@ public void Handle(AlbumGrabbedEvent message)
         public void Handle(AlbumEditedEvent message)
         {
-            BroadcastResourceChange(ModelAction.Updated, message.Album.Id);
+            BroadcastResourceChange(ModelAction.Updated, message.Album.ToResource());
```
See above note regarding artists. Same applies.

Qstick commented on src/Lidarr.Api.V1/Tracks/TrackModuleWithSignalR.cs:
```diff
@@ -107,15 +107,15 @@ public void Handle(TrackImportedEvent message)
             foreach (var track in message.TrackFile.Tracks.Value)
             {
-                BroadcastResourceChange(ModelAction.Updated, track.Id);
+                BroadcastResourceChange(ModelAction.Updated, track.ToResource());
```
Same as artist and album above

ta264 replied:
Is this one necessary? I suspect this change was responsible for a lot of the speedups - fetching track by id causes 4 or 5 separate dB queries per track event.  If you get chance could you try referring and see if your slowdown reappears? I never really managed to replicate locally

Qstick commented on src/NzbDrone.Core/Datastore/Migration/024_clear_media_info.cs:
```diff
@@ -0,0 +1,14 @@
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(024)]
+    public class NewMediaInfoFormat : NzbDroneMigrationBase
```
class name format not consistent with other migrations.

Qstick commented on src/NzbDrone.Core/Extras/Lyrics/ExistingLyricImporter.cs:
```diff
@@ -40,9 +42,18 @@ public override IEnumerable<ExtraFile> ProcessFiles(Artist artist, List<string>
+                    {
+                        FileTrackInfo = Parser.Parser.ParseMusicPath(possibleSubtitleFile),
+                        Artist = artist,
+                        Path = possibleSubtitleFile
```
Yes, previous refactor miss on that one.

ta264 commented on build.sh:
```diff
@@ -254,15 +251,17 @@ PackageTests()
     mkdir $testPackageFolder
 
-    find $sourceFolder -path $testSearchPattern -exec cp -r -u -T "{}" $testPackageFolder \;
+    find . -maxdepth 6 -path $testSearchPattern -exec cp -r "{}" $testPackageFolder \;
```
I forget why that was there - I stuck it in when I did the donate build script years ago.  This was just reverting a commit that broke the macos packaging and so stopped tests running there.  Will take a look.

ta264 replied:
Had a look at this and otherwise you end up flattening the directory structure.  The `find` command gets all the files and directories matching `*.Test/bin/x86/Release/*` and recursively copies them into `_test`.  If you omit the `-maxdepth 6`, as well as the recursively copied directories you get every single matched file copied into `_test` also.  The assumption is that all tests projects are under `src` and output to `src/[something].Test/bin/x86/Release` which is 5 levels deep and so we'd get a recursive copy of everything in that `Release` folder.

So it's a bit of a bodge but it would only cause grief if someone set up a whole new test project in `src/MyRandomTests/Lidarr.Awesome.Test`

ta264 commented on package.json:
```diff
@@ -6,8 +6,9 @@
-    "stylelint": "stylelint frontend/**/*.css --config frontend/.stylelintrc"
+    "eslint-fix": "eslint --fix frontend/** ",
+    "stylelint-linux": "stylelint $(find frontend -name '*.css') --config frontend/.stylelintrc",
+    "stylelint-windows": "stylelint frontend/**/*.css --config frontend/.stylelintrc"
```
That glob only works in `bash` but the yarn thing forces you to use `sh` so it doesn't actually work in the script even though it looks like it works when you copy paste into a terminal

ta264 commented on src/Lidarr.Api.V1/TrackFiles/MediaInfoResource.cs:
```diff
@@ -1,5 +1,6 @@
-using NzbDrone.Core.MediaFiles.MediaInfo;
+using NzbDrone.Core.MediaFiles;
```
These happened because I took out `MediaInfo` and the `MediaInfoModel` class moved namespace

ta264 commented on src/Marr.Data/QGen/QueryBuilder.cs:
```diff
@@ -568,6 +568,23 @@ protected override Expression VisitMethodCall(MethodCallExpression expression)
+            _isJoin = true;
+            MemberInfo rightMember = (rightEntity.Body as MemberExpression).Member;
+
+            foreach (var item in EntGraph)
```
I also haven't really dug into this - the mapper scares me! This is just a copy paste of the function above with a slightly different signature to get one of the joins to compile

ta264 commented on src/NzbDrone.Core.Test/NzbDrone.Core.Test.csproj:
```diff
@@ -594,8 +604,14 @@
+    </ItemGroup>
+    <Copy
+        SourceFiles="@(IdentificationTestCases)"
+        DestinationFolder="$(OutputPath)\Files\Identification\"
```
This just copies into the same `Files` folder that exists for the other tests but doing it this way means you don't have to explicitly name everything to be copied, it just takes the entire directory

ta264 commented on src/NzbDrone.Core/Datastore/BasicRepository.cs:
```diff
@@ -244,6 +245,30 @@ public void SetFields(TModel model, params Expression<Func<TModel, object>>[] pr
+                {
+                    if (model.Id == 0)
+                    {
+                        throw new InvalidOperationException("Can't update model with ID 0");
```
The `using` essentially compiles into a try/catch/finally and makes sure `unitOfWork` gets disposed when the exception throws.

ta264 commented on src/NzbDrone.Core/Datastore/TableMapping.cs:
```diff
@@ -119,40 +120,40 @@ public static void Map()
-                   "WHERE AlbumReleases.Id = {0}",
-                   t.AlbumReleaseId)).SingleOrDefault());
+                .LazyLoad(condition: t => t.AlbumReleaseId > 0, query: (db, t) => db.Query<Artist>()
+                          .Join<Artist, ArtistMetadata>(JoinType.Inner, a => a.Metadata, (a, m) => a.ArtistMetadataId == m.Id)
```
In my testing this made the import quite a bit faster.  They get converted into sqlite joins (which look almost identical to the old queries when I turned on trace query logging) by the mapper but pull back more columns so that the joined objects are all populated instead of needing extra lazyload calls to the database.  I think this really helped with the slowdown @Qstick was seeing

ta264 commented on src/NzbDrone.Core.Test/MediaFiles/TrackImport/Identification/TrackGroupingServiceFixture.cs:
```diff
@@ -0,0 +1,363 @@
+        [TestCase("Various Artists")]
+        [TestCase("Various")]
+        [TestCase("VA")]
+        [TestCase("Unknown")]
```
Good point.  Ive never really looked at a real life various artists release so not sure how they're normally tagged.  I pretty much just copied this from beets.  What would you guys suggest?

Qstick replied:
I think I'm good with the various cases as is for now. May be good to add a test for artists that contain "VA" but should not match like "Va Va Voom". It currently works fine, but a test would be good to ensure its not broken in future.

ta264 commented:
Thanks for taking such a detailed look @danielunderwood.  Will make the changes you suggest when I get a chance.

ta264 commented on src/NzbDrone.Core.Test/Download/CompletedDownloadServiceFixture.cs:
```diff
@@ -162,18 +177,49 @@ public void should_not_process_if_output_path_is_empty()
+                           {
+                               new ImportResult(
+                                   new ImportDecision<LocalTrack>(
+                                       new LocalTrack {Path = @"C:\TestPath\Droned.S01E01.mkv".AsOsAgnostic()})),
```
These were just copy/paste from the other tests.  Loads of the tests have `.mkv` in and it doesn't seem to hurt anything so just thought I'd leave it consistent.

ta264 commented:
I think I've addressed all of @danielunderwood's comments except for which artist tags we should count as Various Artists.  Input on that appreciated.

@danielunderwood's comments on the mapper reminded me that I discovered that as currently set up, if you explicitly `Join` a one-to-many `LazyLoaded` relationship then the items don't get populated and you just get an empty list.  I've fixed that in 811b470 so no-one else has to waste a day discovering the problem again!

danielunderwood commented:
Looks good to me on the code side now unless we want to add braces to that conditional at IEnumerableExtensions.cs:57. I'm not terribly sure what should be concerned as various artists since I haven't seen too many samples. We could parse that case in a special way, but that may be a bit dirty. @Qstick?

ta264 commented:
I think braces are there, it just didn't show your comment as outdated, not sure why.
